test/units/modules/cloud/amazon/placebo_fixtures.py

```diff
@@ -0,0 +1,192 @@
+from nose.plugins.skip import SkipTest
+import pytest
+import os
+
+try:
+    import boto3
+    import botocore
+    import placebo
+except ImportError:
+    raise SkipTest("test_ec2_asg.py requires the `boto3`, `botocore`, and `placebo` modules")
+
+
+"""
+Using Placebo to test modules using boto3:
+
+This is an example test, using the placeboify fixture to test that a module
+will fail if resources it depends on don't exist.
+
+> from placebo_fixtures import placeboify, scratch_vpc
+>
+> def test_create_with_nonexistent_launch_config(placeboify):
+>     connection = placeboify.client('autoscaling')
+>     module = FakeModule('test-asg-created', None, min_size=0, max_size=0, desired_capacity=0)
+>     with pytest.raises(FailJSON) as excinfo:
+>         asg_module.create_autoscaling_group(connection, module)
+>     .... asserts based on module state/exceptions ....
+
+In more advanced cases, use unrecorded resource fixtures to fill in ARNs/IDs of
+things modules depend on, such as:
+
+> def test_create_in_vpc(placeboify, scratch_vpc):
+>     connection = placeboify.client('autoscaling')
+>     module = FakeModule(name='test-asg-created',
+>         min_size=0, max_size=0, desired_capacity=0,
+>         availability_zones=[s['az'] for s in scratch_vpc['subnets']],
+>         vpc_zone_identifier=[s['id'] for s in scratch_vpc['subnets']],
+>     )
+>     ..... so on and so forth ....
+"""
+
+
+@pytest.fixture
+def placeboify(request, monkeypatch):
+    """This fixture puts a recording/replaying harness around `boto3_conn`
+
+    Placeboify patches the `boto3_conn` function in ec2 module_utils to return
+    a boto3 session that in recording or replaying mode, depending on the
+    PLACEBO_RECORD environment variable. Unset PLACEBO_RECORD (the common case
+    for just running tests) will put placebo in replay mode, set PLACEBO_RECORD
+    to any value to turn off replay & operate on real AWS resources.
+
+    The recorded sessions are stored in the test file's directory, under the
+    namespace `placebo_recordings/{testfile name}/{test function name}` to
+    distinguish them.
+    """
+    session = boto3.Session()
+
+    recordings_path = os.path.join(
+        request.fspath.dirname,
+        'placebo_recordings',
+        request.fspath.basename.replace('.py', ''),
+        request.function.func_name
+        # remove the test_ prefix from the function & file name
+    ).replace('test_', '')
+
+    try:
+        # make sure the directory for placebo test recordings is available
+        os.makedirs(recordings_path)
+    except OSError as e:
+        if e.errno != os.errno.EEXIST:
+            raise
+
+    pill = placebo.attach(session, data_path=recordings_path)
+    if os.getenv('PLACEBO_RECORD'):
+        pill.record()
+    else:
+        pill.playback()
+
+    def boto3_middleman_connection(module, conn_type, resource, region, **kwargs):
```

Comment:
Are there guidelines to determine when to make a function inside a function? What is the reasoning for this?